components/service/sync-autofill/src/main/java/mozilla/components/service/sync/autofill/AutofillCrypto.kt

```diff
@@ -0,0 +1,135 @@
+            }
+            KeyGenerationReason.AbnormalState -> {
+                logger.warn("Abnormal state while reading the key")
+            }
+            null, KeyGenerationReason.New -> {
+                // All good! Got either a brand new key or read a valid key.
+            }
+        }
+        return managedKey
+    }
+
+    private fun encrypt(key: ManagedKey, cleartext: String) = encryptString(key.key, cleartext)
+    private fun decrypt(key: ManagedKey, ciphertext: String) = decryptString(key.key, ciphertext)
+
+    private fun getManagedKey(): ManagedKey = synchronized(this) {
+        val encryptedCanaryPhrase = plaintextPrefs.getString(CANARY_PHRASE_CIPHERTEXT_KEY, null)
+        val storedKey = securePrefs.getString(AUTOFILL_KEY)
+
+        return@synchronized when {
+            // We expected the key to be present, and it is.
+            storedKey != null && encryptedCanaryPhrase != null -> {
+                // Make sure that the key is valid.
+                try {
+                    if (CANARY_PHRASE_PLAINTEXT == decryptString(storedKey, encryptedCanaryPhrase)) {
+                        ManagedKey(storedKey)
+                    } else {
+                        ManagedKey(generateAndStoreKey(), KeyGenerationReason.Corrupt)
+                    }
+                } catch (e: ErrorException.JsonError) {
+                    ManagedKey(generateAndStoreKey(), KeyGenerationReason.Corrupt)
+                } catch (e: ErrorException.CryptoError) {
+                    ManagedKey(generateAndStoreKey(), KeyGenerationReason.Corrupt)
+                }
+            }
+
+            // The key is present, but we didn't expect it to be there.
+            storedKey != null && encryptedCanaryPhrase == null -> {
+                // This isn't expected to happen. We can't check this key's validity.
+                ManagedKey(generateAndStoreKey(), KeyGenerationReason.AbnormalState)
+            }
+
+            // We expected the key to be present, but it's gone missing on us.
+            storedKey == null && encryptedCanaryPhrase != null -> {
+                // At this point, we're forced to generate a new key to recover and move forward.
+                // However, that means that any data that was previously encrypted is now unreadable.
+                ManagedKey(generateAndStoreKey(), KeyGenerationReason.Lost)
+            }
+
+            // We didn't expect the key to be present, and it's not.
+            storedKey == null && encryptedCanaryPhrase == null -> {
+                // Normal case when interacting with this class for the first time.
+                ManagedKey(generateAndStoreKey(), KeyGenerationReason.New)
+            }
+
+            else -> throw AutofillCryptoException.IllegalState()
+        }
+    }
+
+    private fun generateAndStoreKey(): String {
+        return createKey().also { newKey ->
+            // To consider: should this be a non-destructive operation, just in case?
```

Comment:
yeah, I've thought about this - a transient error here would suck (eg, if it might work after the device restarts) but that seems tricky to rationalize about at this layer.